File: src/dcs_mission_creator/core/lua/intercept.lua
-- dcs_mission_creator: intercept, shadow and escort an intruder out of a zone.
--
-- Rendered by `core/intercept.py` into a mission-start DoScript. Placeholders
-- (named here without their underscores so this comment is not substituted):
-- TRACKS is one row per intruder group, ZONE the boundary polygon as `{x=, y=}`
-- rows, SIDE the players' coalition, INTERCEPT M / HOLD S how close a player
-- has to be and for how long, ESCORT M / LAPSE S how far a player may drift and
-- for how long before an escorted track turns back in, TICK the poll period,
-- CONTROLLER the group whose radio every call goes out on (nil for none),
-- DEFENDERS the friendly AI groups committed on a hostile, the three FLAG
-- numbers, and TRACE.
--
-- The decisions are the Python module's docstring; what is here is the state
-- machine, one per track:
--
--   pending   late-activated, not yet flying
--   loose     inside the zone on its own route, nobody alongside
--   escorted  a player held a position on it, it is flying home
--   hostile   it turned on the escort; the defenders are committed
--   resolved  out of the zone, or dead
do
  local tracks = {
__TRACKS__
  }
  local zone = {
__ZONE__
  }
  local side = __SIDE__
  local interceptM = __INTERCEPT_M__
  local holdS = __HOLD_S__
  local escortM = __ESCORT_M__
  local lapseS = __LAPSE_S__
  local tick = __TICK__
  local controller = __CONTROLLER__
  local defenders = {__DEFENDERS__}
  local flagDone, flagFoul, flagUnchallenged = __FLAG_DONE__, __FLAG_FOUL__, __FLAG_UNCHALLENGED__
  local trace = __TRACE__

  local byGroup = {}
  for _, t in ipairs(tracks) do
    t.state = "pending"
    t.near = 0
    t.away = 0
    t.epoch = 0
    byGroup[t.group] = t
  end

  local function log(t, fmt, ...)
    if trace then
      env.info(string.format("INTERCEPT/%s t+%.0fs: ", t.label, timer.getTime())
        .. string.format(fmt, ...))
    end
  end

  -- Calls go out one at a time, and only while the controller is alive to make
  -- them: a picture nobody is left to transmit is not transmitted.
  local queue, draining = {}, false
  local function controllerAlive()
    if controller == nil then return true end
    local g = Group.getByName(controller)
    if g == nil or not g:isExist() then return false end
    for _, u in ipairs(g:getUnits()) do
      if u:isExist() and u:getLife() > 0 then return true end
    end
    return false
  end
  local function playNext(_, time)
    local item = table.remove(queue, 1)
    if item == nil then
      draining = false
      return nil
    end
    if item.text then trigger.action.outTextForCoalition(side, item.text, 15) end
    if item.sound then trigger.action.outSoundForCoalition(side, item.sound) end
    return time + 8
  end
  local function say(t, key)
    local call = t.calls[key]
    if call == nil or not controllerAlive() then return end
    queue[#queue + 1] = call
    if not draining then
      draining = true
      timer.scheduleFunction(playNext, {}, timer.getTime() + 0.1)
    end
  end

  -- Ray casting against the boundary, in DCS x (north) / z (east).
  local function inside(p)
    local hit = false
    local j = #zone
    for i = 1, #zone do
      local a, b = zone[i], zone[j]
      if ((a.y > p.z) ~= (b.y > p.z))
        and (p.x < (b.x - a.x) * (p.z - a.y) / (b.y - a.y) + a.x) then
        hit = not hit
      end
      j = i
    end
    return hit
  end

  local function live(t)
    local g = Group.getByName(t.group)
    if g == nil or not g:isExist() then return {} end
    local out = {}
    for _, u in ipairs(g:getUnits()) do
      if u:isExist() and u:isActive() and u:getLife() > 0 then out[#out + 1] = u end
    end
    return out
  end

  local function nearestPlayer(units)
    local best, bestD = nil, math.huge
    for _, pu in ipairs(coalition.getPlayers(side)) do
      if pu:isExist() and pu:inAir() then
        local pp = pu:getPoint()
        for _, u in ipairs(units) do
          local up = u:getPoint()
          local dx, dy, dz = pp.x - up.x, pp.y - up.y, pp.z - up.z
          local d = math.sqrt(dx * dx + dy * dy + dz * dz)
          if d < bestD then best, bestD = pu, d end
        end
      end
    end
    return best, bestD
  end

  local function lead(units)
    return units[1]:getPoint()
  end

  local function turningPoint(p, alt, speed, tasks)
    return {
      type = "Turning Point", action = "Turning Point",
      x = p.x, y = p.y, alt = alt, alt_type = "BARO", speed = speed,
      task = {id = "ComboTask", params = {tasks = tasks or {}}},
    }
  end

  local function homeRoute(t, from, extra)
    return {
      turningPoint({x = from.x, y = from.z}, t.alt, t.speed, extra),
      turningPoint(t.exit, t.alt, t.speed),
      {
        type = "Land", action = "Landing", airdromeId = t.airdrome,
        x = t.home.x, y = t.home.y, alt = 0, alt_type = "BARO", speed = t.speed,
      },
    }
  end

  local function setMission(t, points)
    local g = Group.getByName(t.group)
    if g == nil then return end
    g:getController():setTask({id = "Mission", params = {route = {points = points}}})
  end

  local function setRoe(groupName, value)
    local g = Group.getByName(groupName)
    if g == nil or not g:isExist() then return end
    g:getController():setOption(AI.Option.Air.id.ROE, value)
  end

  -- The only entropy a mission script has is the player: a millisecond clock
  -- read at the moment somebody arrives alongside is different every sortie,
  -- where the stock generator replays the same stream from mission start.
  local function reseed(p)
    math.randomseed(math.floor(timer.getTime() * 1000) + math.floor(math.abs(p.x) % 997))
    math.random(); math.random()
  end

  local function sendHome(t, units)
    setRoe(t.group, AI.Option.Air.val.ROE.WEAPON_HOLD)
    setMission(t, homeRoute(t, lead(units)))
  end

  local function turnHostile(t, units)
    t.state = "hostile"
    local escort = nearestPlayer(units)
    local via = escort and escort:getPoint() or lead(units)
    setRoe(t.group, AI.Option.Air.val.ROE.WEAPON_FREE)
    local g = Group.getByName(t.group)
    g:getController():setOption(AI.Option.Air.id.REACTION_ON_THREAT,
      AI.Option.Air.val.REACTION_ON_THREAT.EVADE_FIRE)
    local engage = {id = "EngageTargets", params = {
      targetTypes = {"Air"}, priority = 0, maxDist = 80000,
    }}
    local points = homeRoute(t, lead(units), {engage})
    table.insert(points, 2, turningPoint({x = via.x, y = via.z}, t.alt, t.speed * 1.2))
    setMission(t, points)
    -- The defenders are held on OPEN_FIRE, which in DCS is "only what you are
    -- tasked against", so committing them on this group cannot turn them on a
    -- compliant track two miles away.
    for _, name in ipairs(defenders) do
      local d = Group.getByName(name)
      if d and d:isExist() then
        d:getController():setOption(AI.Option.Air.id.ROE, AI.Option.Air.val.ROE.OPEN_FIRE)
        d:getController():pushTask({id = "AttackGroup", params = {groupId = g:getID()}})
      end
    end
    log(t, "hostile; %d defender group(s) committed", #defenders)
    say(t, "hostile")
  end

  local function scheduleRoll(t, units)
    reseed(lead(units))
    local roll = math.random()
    if roll >= t.p then
      log(t, "roll %.2f >= %.2f, stays compliant", roll, t.p)
      return
    end
    local wait = t.after[1] + (math.random() + math.random()) * 0.5 * (t.after[2] - t.after[1])
    local epoch = t.epoch
    log(t, "roll %.2f < %.2f, turns in %.0fs if still escorted", roll, t.p, wait)
    timer.scheduleFunction(function()
      if t.state ~= "escorted" or t.epoch ~= epoch then return nil end
      local now = live(t)
      if #now == 0 or not inside(now[1]:getPoint()) then return nil end
      turnHostile(t, now)
      return nil
    end, {}, timer.getTime() + wait)
  end

  local function resolve(t, how)
    t.state = "resolved"
    t.how = how
    log(t, "resolved: %s", how)
    for _, other in ipairs(tracks) do
      if other.state ~= "resolved" then return end
    end
    trigger.action.setUserFlag(flagDone, 1)
  end

  local function step(t, now)
    if t.state == "resolved" then return end
    local units = live(t)
    if t.state == "pending" then
      if #units == 0 then return end
      t.state = "loose"
      t.since = now
      log(t, "active")
      return
    end
    if #units == 0 then
      say(t, t.state == "hostile" and "splashed" or "down")
      resolve(t, t.state == "hostile" and "splashed" or "down")
      return
    end
    local p = lead(units)
    local _, d = nearestPlayer(units)

    if t.state == "loose" then
      if t.unchallenged then
        if not inside(p) then
          resolve(t, "left unchallenged")
        end
        return
      end
      t.near = (d <= interceptM) and (t.near + tick) or 0
      if t.near >= holdS then
        t.state = "escorted"
        t.away, t.epoch = 0, t.epoch + 1
        log(t, "intercepted at %.0f m", d)
        sendHome(t, units)
        say(t, "intercepted")
        scheduleRoll(t, units)
      elseif now - t.since >= t.deadline then
        t.unchallenged = true
        trigger.action.setUserFlag(flagUnchallenged, 1)
        log(t, "deadline passed with nobody alongside")
        setMission(t, homeRoute(t, p))
        say(t, "unchallenged")
      end
      return
    end

    if not inside(p) then
      say(t, t.state == "hostile" and "fled" or "cleared")
      resolve(t, t.state == "hostile" and "fled" or "escorted out")
      return
    end

    if t.state == "escorted" then
      t.away = (d > escortM) and (t.away + tick) or 0
      if t.away >= lapseS then
        t.state, t.near, t.epoch = "loose", 0, t.epoch + 1
        log(t, "escort lapsed at %.0f m, back to the orbit", d)
        setMission(t, {
          turningPoint({x = p.x, y = p.z}, t.alt, t.speed),
          turningPoint(t.orbit, t.alt, t.speed, {{id = "Orbit", params = {
            pattern = "Circle", altitude = t.alt, speed = t.speed,
          }}}),
        })
        say(t, "lapsed")
      end
    end
  end

  local function poll(_, time)
    for _, t in ipairs(tracks) do
      local ok, err = pcall(step, t, time)
      if not ok then env.error("INTERCEPT/" .. t.label .. ": " .. tostring(err)) end
    end
    return time + tick
  end

  -- A weapon on a track that never turned is the incident the ROE exists to
  -- prevent; one hit is enough, whoever's missile it was on our side.
  local handler = {}
  function handler:onEvent(event)
    if event.id ~= world.event.S_EVENT_HIT or event.target == nil then return end
    if event.initiator == nil or not event.initiator.getCoalition then return end
    if event.initiator:getCoalition() ~= side then return end
    local ok, group = pcall(function() return event.target:getGroup() end)
    if not ok or group == nil then return end
    local t = byGroup[group:getName()]
    if t == nil or t.state == "hostile" or t.fouled then return end
    t.fouled = true
    trigger.action.setUserFlag(flagFoul, 1)
    log(t, "hit while not hostile")
    say(t, "foul")
  end
  world.addEventHandler(handler)

  timer.scheduleFunction(poll, {}, timer.getTime() + 1)
end

File: src/dcs_mission_creator/core/intercept.py
"""Intercept an intruder, shadow it, and escort it out of a zone (project-owned).

DCS has no model of an aircraft that is *not yet* an enemy. Two coalitions are
at war from the first frame: a red Su-27 over a blue CAP is a target, and a blue
CAP over a red Su-27 shoots it. Air policing is the opposite problem — the
intruder is a coalition enemy that nobody is cleared to fire on until it does
something — and the mission editor reaches none of the three things it needs:

- **"Intercepted" is a formation state, not a trigger.** No condition says "a
  player has been sitting on this aircraft's wing". The nearest is a moving
  zone, which cannot express "for long enough to be seen" and cannot re-arm.
- **Compliance is a change of route.** An intercepted aircraft turns for its
  own field, which is a new `Mission` task built from where it is *now*.
- **The flip is a coin the pilot cannot see.** It has to be rolled at run
  time: pydcs' RNG is seeded from the mission slug (`MissionBuilder._seed_rng`),
  so a roll made at build time would be the same roll on every sortie flown on
  that `.miz`.

So all three live in `core/lua/intercept.lua`, one state machine per track
(`pending → loose → escorted → hostile → resolved`, with `escorted → loose`
when the escort drifts off). What is here is the table that feeds it and the
editor-side half that has to be true before the script runs.

**Why the defenders are held on `OPEN_FIRE` rather than `WEAPON_HOLD` alone.**
A friendly CAP is the obvious backstop for a track that turns hostile, and the
obvious way to commit it — weapons free — commits it on *every* red aircraft in
range, including the compliant pair its own lead is escorting. DCS's
`OPEN_FIRE` is "engage only what you are tasked against", so on the flip the
script sets that and pushes an `AttackGroup` on the one group that turned. To
make it true the defender must carry no en-route engage task of its own — the
`CAPTaskAction` `Mission._load_tasks` adds for a CAP main task is exactly that —
so `arm_intercepts` strips them.

**Why a compliant track is resolved on leaving the zone, and not on reaching
home.** The mission's claim is sovereignty over an airspace; what happens north
of the line is somebody else's. It is also what makes the escort finite.

**Why a lapse re-arms the roll.** A track left alone turns back to its orbit
and has to be intercepted again, and the second intercept rolls again. A player
who lets them go pays for it twice — in time, and in the chance of a flip.

**Why the entropy comes from the player.** DCS's scripting `math.random`
replays one stream from mission start. The only thing that differs between two
sorties on the same file is the human, so the generator is reseeded at each
intercept off the millisecond clock and the intruder's position.

**Why one hit ends the policing, whoever fired.** The foul flag is on any hit
by the players' coalition on a track that has not turned, AI defenders
included; a compliant aircraft shot down is the one outcome the whole ROE
exists to prevent, and it is not less of an incident because a wingman's
missile did it.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Optional, Sequence

import structlog
from dcs import task, triggers

from dcs_mission_creator.core import lua

if TYPE_CHECKING:
    from dcs.mapping import Point
    from dcs.mission import Mission
    from dcs.terrain.terrain import Airport
    from dcs.unitgroup import FlyingGroup

    from dcs_mission_creator.core.tts import VoiceSynth

log = structlog.get_logger(__name__)

__all__ = ["Calls", "Flags", "Track", "arm_intercepts", "inside"]

_SCRIPT = "intercept.lua"
_SIDE = {"blue": "coalition.side.BLUE", "red": "coalition.side.RED"}
_HOSTILE_ONLY = frozenset({"hostile", "splashed", "fled"})
#: En-route tasks that let a flight pick its own targets. `CAPTaskAction` is the
#: one `Mission._load_tasks` adds for a CAP main task; the other two are what a
#: mission would reach for by hand.
_SELF_ENGAGING = (task.CAPTaskAction, task.EngageTargets, task.EngageTargetsInZone)

#: A player this close, 3D, counts as alongside. One nautical mile is the
#: ICAO interception procedure's own "visual, abeam the cockpit" distance.
INTERCEPT_M = 1_852.0
#: How long alongside before the intruder is taken to have seen the escort.
HOLD_S = 20.0
#: How far an escort may drift from a compliant track before it notices.
ESCORT_M = 18_520.0
#: How long it has to be unwatched before it turns back in.
LAPSE_S = 90.0
#: Poll period. Two seconds is under a tenth of `HOLD_S`, so the hold is
#: resolved to within one tick without the script walking every unit per frame.
TICK_S = 2.0


@dataclass(frozen=True)
class Calls:
    """What the controller says about a track, as `{label}` templates.

    `None` keeps a call silent. Each one is spoken and printed from Lua with the
    same string (`VoiceSynth.register`), the `core/triggers.py` rule applied to
    a call a trigger cannot make because it can happen more than once.
    """

    intercepted: Optional[str] = None
    hostile: Optional[str] = None
    splashed: Optional[str] = None
    down: Optional[str] = None
    cleared: Optional[str] = None
    fled: Optional[str] = None
    lapsed: Optional[str] = None
    unchallenged: Optional[str] = None
    foul: Optional[str] = None

    def items(self) -> list[tuple[str, str]]:
        return [(k, v) for k, v in vars(self).items() if v is not None]


@dataclass(frozen=True)
class Track:
    """One intruder group and everything it does once intercepted.

    `orbit` is where it loiters inside the zone and where it goes back to on a
    lapse; `exit` is the first point of its way home and must be outside the
    zone; `home` is the field it lands at. `hostile_probability` is rolled at
    each intercept, and a flip happens `hostile_after_s` later, drawn
    triangularly, only if the track is still escorted and still inside.
    `deadline_s` is how long it loiters, counted from activation, before it
    leaves on its own with its job done. Speeds are km/h, like every pydcs
    speed; altitudes metres.
    """

    group: FlyingGroup
    label: str
    orbit: Point
    exit: Point
    home: Airport
    altitude_m: float
    speed_kph: float
    hostile_probability: float
    hostile_after_s: tuple[float, float] = (45.0, 240.0)
    deadline_s: float = 1_500.0
    calls: Calls = field(default_factory=Calls)


@dataclass(frozen=True)
class Flags:
    """User flags the script sets; the mission writes its outcome calls on them.

    `done` — every track resolved (out of the zone or dead). `foul` — a track
    that had not turned was hit by our side. `unchallenged` — at least one track
    loitered its full time with nobody alongside.
    """

    done: int
    foul: int
    unchallenged: int


def inside(p: Point, zone: Sequence[Point]) -> bool:
    """Ray-casting point-in-polygon, the same test the Lua runs."""
    hit = False
    j = len(zone) - 1
    for i in range(len(zone)):
        a, b = zone[i], zone[j]
        if (a.y > p.y) != (b.y > p.y) and p.x < (b.x - a.x) * (p.y - a.y) / (
            b.y - a.y
        ) + a.x:
            hit = not hit
        j = i
    return hit


def _validate(tracks: Sequence[Track], zone: Sequence[Point]) -> None:
    if len(zone) < 3:
        raise ValueError("an intercept zone needs at least three vertices")
    if not tracks:
        raise ValueError("arm_intercepts needs at least one track")
    for t in tracks:
        if not 0.0 <= t.hostile_probability <= 1.0:
            raise ValueError(f"{t.label}: hostile_probability must be in [0, 1]")
        lo, hi = t.hostile_after_s
        if not 0.0 <= lo <= hi:
            raise ValueError(f"{t.label}: hostile_after_s must be (lo, hi), lo <= hi")
        if not inside(t.orbit, zone):
            raise ValueError(f"{t.label}: its orbit is outside the zone")
        if inside(t.exit, zone):
            # A track whose way home starts inside the zone may never leave it
            # on the compliant route, and the mission never resolves.
            raise ValueError(f"{t.label}: its exit point is inside the zone")
        if not t.group.units:
            raise ValueError(f"{t.label}: {t.group.name} has no units")


def _hold_fire(group: FlyingGroup) -> None:
    """Weapons hold from the spawn waypoint, appended last so it wins."""
    group.points[0].tasks.append(task.OptROE(task.OptROE.Values.WeaponHold))


def _commit_only_when_told(group: FlyingGroup) -> None:
    """A defender engages nothing on its own — see the module docstring."""
    for point in group.points:
        point.tasks = [t for t in point.tasks if not isinstance(t, _SELF_ENGAGING)]
    _hold_fire(group)


def _xy(p: Point) -> str:
    return f"{{x={p.x:.1f}, y={p.y:.1f}}}"


def _row(t: Track, calls: dict[str, tuple[str, Optional[str]]]) -> str:
    call_rows = ", ".join(
        f"{key}={{text={lua.quote(text)}, sound={lua.quote(sound)}}}"
        for key, (text, sound) in calls.items()
    )
    return (
        "    {{group={group}, label={label}, p={p:.3f}, after={{{lo:.1f}, {hi:.1f}}}, "
        "deadline={deadline:.1f}, alt={alt:.1f}, speed={speed:.2f}, orbit={orbit}, "
        "exit={exit}, home={home}, airdrome={airdrome}, calls={{{calls}}}}},".format(
            group=lua.quote(t.group.name),
            label=lua.quote(t.label),
            p=t.hostile_probability,
            lo=t.hostile_after_s[0],
            hi=t.hostile_after_s[1],
            deadline=t.deadline_s,
            alt=t.altitude_m,
            speed=t.speed_kph / 3.6,
            orbit=_xy(t.orbit),
            exit=_xy(t.exit),
            home=_xy(t.home.position),
            airdrome=int(t.home.id),
            calls=call_rows,
        )
    )


def arm_intercepts(
    m: Mission,
    tracks: Sequence[Track],
    *,
    zone: Sequence[Point],
    defenders: Sequence[FlyingGroup] = (),
    controller: Optional[FlyingGroup] = None,
    voice: Optional[VoiceSynth] = None,
    coalition: str = "blue",
    intercept_m: float = INTERCEPT_M,
    hold_s: float = HOLD_S,
    escort_m: float = ESCORT_M,
    lapse_s: float = LAPSE_S,
    flag_base: int = 900,
    trace: bool = True,
) -> Flags:
    """Arm the intercept state machine for `tracks` inside `zone`.

    Sets every track and every defender to weapons hold from spawn, strips the
    defenders' own engage tasks, and adds one mission-start `DoScript`.
    `controller` is whose radio the calls go out on — the AWACS — and they stop
    when it is dead. `trace` writes each decision to `dcs.log` under
    `INTERCEPT/<label>`, and nothing to the screen.

    Returns the three flags the mission writes its outcome calls on, numbered
    from `flag_base`.
    """
    if coalition not in _SIDE:
        raise ValueError(f"coalition must be blue/red, got {coalition!r}")
    _validate(tracks, zone)
    for t in tracks:
        _hold_fire(t.group)
    for d in defenders:
        _commit_only_when_told(d)

    rows: list[str] = []
    for t in tracks:
        calls: dict[str, tuple[str, Optional[str]]] = {}
        for key, template in t.calls.items():
            if key in _HOSTILE_ONLY and t.hostile_probability == 0.0:
                continue  # a call this track cannot make is not worth a render
            text = template.format(label=t.label)
            calls[key] = (text, voice.register(m, text) if voice else None)
        rows.append(_row(t, calls))

    flags = Flags(done=flag_base, foul=flag_base + 1, unchallenged=flag_base + 2)
    script = lua.render(
        _SCRIPT,
        TRACKS="\n".join(rows),
        ZONE="\n".join(f"    {_xy(p)}," for p in zone),
        SIDE=_SIDE[coalition],
        INTERCEPT_M=f"{intercept_m:.1f}",
        HOLD_S=f"{hold_s:.1f}",
        ESCORT_M=f"{escort_m:.1f}",
        LAPSE_S=f"{lapse_s:.1f}",
        TICK=f"{TICK_S:.1f}",
        CONTROLLER=lua.quote(controller.name if controller else None),
        DEFENDERS=", ".join(lua.quote(d.name) for d in defenders),
        FLAG_DONE=str(flags.done),
        FLAG_FOUL=str(flags.foul),
        FLAG_UNCHALLENGED=str(flags.unchallenged),
        TRACE="true" if trace else "false",
    )
    rule = triggers.TriggerStart(comment="Intercept — escort-out state machine")
    rule.add_action(lua.InlineDoScript(script))
    m.triggerrules.triggers.append(rule)
    log.debug(
        "armed intercepts",
        tracks=[t.label for t in tracks],
        defenders=[d.name for d in defenders],
        flags=flags,
    )
    return flags
